Backport modern section change logic from multires

ds-sloth opened the pull request:
Partial port of the modern section change logic from multires. Does not contain the logic for `vScreen` position (`ScreenTop` / `ScreenLeft`) or for the "quick section change" used for star exits in the Yoshi's Archipelago episode.

This code is all well-tested and should be ready for immediate merge, but I'd like @Wohlstand's quick review before merging.

At 800x600, there will be two visible effects. First, bug #43 will not occur. Second, 2P camera merges / splits will be animated.

Fixes #43.

<img src="https://github.com/Wohlstand/TheXTech/assets/72112344/60f39681-dd7c-4dc2-985a-49dffbc8d25d" width="400" height="300">

<img src="https://github.com/Wohlstand/TheXTech/assets/72112344/516de216-419e-4390-a103-624c7b22957e" width="400" height="300">

[section resize.lvl.zip](https://github.com/Wohlstand/TheXTech/files/12196216/section.resize.lvl.zip)

Wohlstand commented on src/layers.cpp:
```diff
@@ -822,19 +840,132 @@ void ProcEvent(eventindex_t index, bool NoEffect)
+                                double maxShiftX = vScreen[1].Width / 4;
+                                double maxShiftY = vScreen[1].Height / 4;
+                                if(maxShiftX > 200)
+                                    maxShiftX = 200;
+                                if(maxShiftY > 150)
```
Empty line please between if conditions except `else if`.

Wohlstand commented on src/layers.cpp:
```diff
@@ -822,19 +840,132 @@ void ProcEvent(eventindex_t index, bool NoEffect)
+                                qScreenLoc[1].Y += maxShiftY * yProp;
+
+                                // restrict to old level bounds
+                                if(-qScreenLoc[1].X < level[B].X)
+                                    qScreenLoc[1].X = -level[B].X;
+                                if(-qScreenLoc[1].X + vScreen[1].Width > level[B].Width)
+                                    qScreenLoc[1].X = -(level[B].Width - vScreen[1].Width);
+                                if(-qScreenLoc[1].Y < level[B].Y)
+                                    qScreenLoc[1].Y = -level[B].Y;
+                                if(-qScreenLoc[1].Y + vScreen[1].Height > level[B].Height)
+                                    qScreenLoc[1].Y = -(level[B].Height - vScreen[1].Height);
```
Here too:

> Empty line please between if conditions except `else if`.

Wohlstand commented on src/layers.cpp:
```diff
@@ -851,14 +982,16 @@ void ProcEvent(eventindex_t index, bool NoEffect)
+                            // restrict to old level bounds
+                            if(-qScreenLoc[1].X < level[B].X)
+                                qScreenLoc[1].X = -level[B].X;
+                            if(-qScreenLoc[1].X + ScreenW /*FrmMain.ScaleWidth*/ > level[B].Width)
+                                qScreenLoc[1].X = -(level[B].Width - ScreenW);
+                            if(-qScreenLoc[1].Y < level[B].Y)
+                                qScreenLoc[1].Y = -level[B].Y;
```
And here:

> Empty line please between if conditions except `else if`.

Wohlstand commented on src/layers.cpp:
```diff
@@ -822,19 +840,132 @@ void ProcEvent(eventindex_t index, bool NoEffect)
+                                double xProp = tX * tX / dSquare;
+                                double yProp = tY * tY / dSquare;
+
+                                if(tX < 0)
+                                    xProp *= -1;
+                                if(tY < 0)
+                                    yProp *= -1;
```
And here too:

> Empty line please between if conditions except `else if`.

Wohlstand commented on src/layers.cpp:
```diff
@@ -822,19 +840,132 @@ void ProcEvent(eventindex_t index, bool NoEffect)
+                            qScreenLoc[1] = vScreen[1];
+
+                            // set the second qScreen if possible
+                            if(vScreen[2].Visible)
+                            {
+                                qScreenLoc[2] = vScreen[2];
+                            }
```
I guess, brackets here are pointless, just one single line will be fine.

Wohlstand reviewed (changes requested):
Looks like fine, if not count code style quibbles. :)

ds-sloth commented:
Fixed your style notes, thanks. Also added some additional comments. Note that I did add brackets around one single-line `else` block to balance with the multiline `if` block before it.

Wohlstand reviewed (approved):
I quickly checked the work of the thing, and seems works just fine!
Probably, if somebody also test the work of the thing, could give one another note. I'll merge this.
